src/dataflow/operators/unordered_input.rs

```diff
@@ -0,0 +1,123 @@
+impl<G: Scope> UnorderedInput<G> for G {
+    fn new_unordered_input<D:Data>(&mut self) -> ((UnorderedHandle<G, D>, Capability<G::Timestamp>), Stream<G, D>) {
+
+        let (output, registrar) = Tee::<G::Timestamp, D>::new();
+        let internal = Rc::new(RefCell::new(CountMap::new()));
+        let produced = Rc::new(RefCell::new(CountMap::new()));
+        let cap = mint_capability(Default::default(), internal.clone());
+        let helper = UnorderedHandle::new(PushCounter::new(output, produced.clone()));
+        let peers = self.peers();
+
+        let index = self.add_operator(UnorderedOperator {
+            internal: internal.clone(),
+            produced: produced.clone(),
+            peers:    peers,
+        });
+
+        return ((helper, cap), Stream::new(Source { index: index, port: 0 }, registrar, self.clone()));
+    }
+}
+
+struct UnorderedOperator<T:Timestamp> {
+    internal:   Rc<RefCell<CountMap<T>>>,
+    produced:   Rc<RefCell<CountMap<T>>>,
+    peers:     usize,
+}
+
+impl<T:Timestamp> Operate<T> for UnorderedOperator<T> {
+    fn name(&self) -> String { "Input".to_owned() }
+    fn inputs(&self) -> usize { 0 }
+    fn outputs(&self) -> usize { 1 }
+
+    fn get_internal_summary(&mut self) -> (Vec<Vec<Antichain<<T as Timestamp>::Summary>>>,
+                                           Vec<CountMap<T>>) {
+        let mut internal = CountMap::new();
+        // augment the counts for each reserved capability.
+        for &(ref time, count) in self.internal.borrow().iter() {
+            internal.update(time, count * (self.peers as i64 - 1));
+        }
+
+        // drain the changes to empty out, and complete the counts for internal.
+        self.internal.borrow_mut().drain_into(&mut internal);
+        (Vec::new(), vec![internal])
+    }
+
+    fn pull_internal_progress(&mut self,_consumed: &mut [CountMap<T>],
+                                         internal: &mut [CountMap<T>],
+                                         produced: &mut [CountMap<T>]) -> bool
+    {
+        self.produced.borrow_mut().drain_into(&mut produced[0]);
+        self.internal.borrow_mut().drain_into(&mut internal[0]);
+        return false;
+    }
+
+    fn notify_me(&self) -> bool { false }
+}
+
+
+/// A handle to an input `Stream`, used to introduce data to a timely dataflow computation.
+pub struct UnorderedHandle<G: Scope, D: Data> {
+    buffer: PushBuffer<G::Timestamp, D, PushCounter<G::Timestamp, D, Tee<G::Timestamp, D>>>,
+}
+
+impl<G: Scope, D: Data> UnorderedHandle<G, D> {
+    fn new(pusher: PushCounter<G::Timestamp, D, Tee<G::Timestamp, D>>) -> UnorderedHandle<G, D> {
+        UnorderedHandle {
+            buffer: PushBuffer::new(pusher),
+        }
+    }
+
+    pub fn session<'b>(&'b mut self, cap: &Capability<G::Timestamp>) -> Session<'b, G::Timestamp, D, PushCounter<G::Timestamp, D, Tee<G::Timestamp, D>>> {
+        self.buffer.session(cap)
+    }
+
+    pub fn flush(&mut self) {
+        self.buffer.cease();
+    }
+}
```

Comment:
It seems like `UnorderedHandle` has (at the moment) just two interesting methods: `session` and `flush`, which just call through to `PushBuffer`'s `session` and `cease` methods. Would it reduce the cognitive overhead at all to remove this wrapper and just return a `PushBuffer`? Or does the wrapper add something else (e.g. a much shorter name to type out :)), now or in the future (e.g. the possible `Drop` impl)?

Alternately, should whatever pleasant things this type does be integrated into `PushBuffer`?